Windows MFTScan: Memory usage performance fixes

dgmcdona opened the pull request:
This fixes a major memory performance issue in the `MFTScan` plugins. We had been struggling with large memory consumption in these plugins before, and I assumed that the issue was with the `TreeGrid` implementation. There ended up being a larger issue within the actual plugin, which was unconverted `object.String` objects being stored in the `record_map` that is populated and passed around to various classmethods. This resulted in a map that was far larger than it needed to be, since those values would only ever be converted to strings for the treegrid anyway. I discovered this by accident when implementing a simple SQLite cache as a replacement for the map - watching the SQLite database, I never saw it grow to more than a few dozen megabytes. 

```
[ins] In [5]: pympler.asizeof.asizeof(rec_name)
Out[5]: 312648

[ins] In [6]: pympler.asizeof.asizeof(str(rec_name))
Out[6]: 64
```

This PR:
- Properly converts values to strings before storing them in the map
- Fixes type hints in the MFT extension classes to reflect that they return `objects.String`, not `str`.
- Updates type hints throughout the MFTScan plugin classes for clarity.
- Reimplements the map as a typed `DefaultDict` for simplicity
- Adds a simple `MFTRecord` class that uses `__slots__` in place of the lists that were being used as map values before. This should result in some more (small) performance gains, and help with readability since values are no longer accessed by index.

You can see how much better it performs on an 8GB sample here - `sys` time that was presumably spent swapping dropped from ~47 seconds to ~8 seconds:
```
~/volatility3 (fix_tests|✚2) $ time vol -f ~/memory_samples/data.lime -r jsonl windows.mftscan.ResidentData > /dev/null
Volatility 3 Framework 2.26.0
Progress:  100.00		PDB scanning finished                  
________________________________________________________
Executed in  400.14 secs    fish           external
   usr time  340.85 secs    0.32 millis  340.85 secs
   sys time   47.16 secs    1.46 millis   47.16 secs
 
~/volatility3 (fix_tests|●1) $ time vol -f ~/memory_samples/data.lime -r jsonl windows.mftscan.ResidentData > /dev/null
Volatility 3 Framework 2.26.0
Progress:  100.00		PDB scanning finished                  
________________________________________________________
Executed in  321.45 secs    fish           external
   usr time  314.03 secs    0.29 millis  314.03 secs
   sys time    8.18 secs    1.21 millis    8.18 secs
```

closes #1725

dgmcdona committed:
MFT Extensions: Fix type hints

These type hints are a bit misleading, and have been updated to reflect
their real return type.

dgmcdona committed:
Windows MFTScan Plugins: Performance fixes

There was a subtle issue that was causing substantial performance issues
in the MFTScan plugins. The `record_map` was purportedly of type
`Dict[str, Tuple[int, str, int]]`, but in reality, the second member was
a list, and its `str` item was actually being populated with unprocessed
values from method calls on the MFT extension classes, which actually
return `object.String`. These objects are substantially larger than
basic `str` types:

```
[ins] In [5]: pympler.asizeof.asizeof(rec_name)
Out[5]: 312648

[ins] In [6]: pympler.asizeof.asizeof(str(rec_name))
Out[6]: 64
```

This caused this dictionary to grow in size to several gigabytes on
larger samples, resulting in thrashing and OOM errors.

ikelos commented on volatility3/framework/plugins/windows/mftscan.py:
```diff
@@ -98,23 +120,24 @@ def enumerate_mft_records(
         # get each of the individual Field Sets
-        mft_object = symbol_table + constants.BANG + "MFT_ENTRY"
-        attribute_object = symbol_table + constants.BANG + "ATTRIBUTE"
+        mft_object_typ_name = symbol_table + constants.BANG + "MFT_ENTRY"
```
Any reason these are missing the `e`?

dgmcdona replied:
Just a reflex of leaving off the 'e' to avoid accidentally shadowing the builtin, I'll add them in here though.

ikelos commented on volatility3/framework/plugins/windows/mftscan.py:
```diff
@@ -98,23 +120,24 @@ def enumerate_mft_records(
+        attribute_object_typ_name = symbol_table + constants.BANG + "ATTRIBUTE"
 
-        record_map = {}
+        record_map: DefaultDict[str, MFTRecord] = DefaultDict(MFTRecord)
```
I'm a little concerned about the use of `DefaultDict` here?  When are we creating blank `MFTRecords`, and does that not mean we need more code to cope with them if we ever look them up?

ikelos commented on volatility3/framework/plugins/windows/mftscan.py:
```diff
@@ -17,6 +23,22 @@
 vollog = logging.getLogger(__name__)
 
 
+class MFTRecord:
```
This feels like it could be `namedtuple`, rather than a class?  `__slots__` helps reduce its memory footprint, but not being a class would help it more I think?

dgmcdona replied:
Surprisingly, a class with `__slots__` seems to use less memory than even a simple tuple:

```
[ins] In [4]: class MFTRecord:
         ...:     # TODO: Change to dataclass with (slots=True) if/when we move minimum
         ...:     # Python version up to 3.10
         ...:     __slots__ = ["record_name", "data_count", "offset"]
         ...: 
         ...:     def __init__(
         ...:         self,
         ...:         record_name: Optional[str] = None,
         ...:         data_count: int = 0,
         ...:         offset: Optional[int] = None,
         ...:     ):
         ...:         self.record_name = record_name
         ...:         self.data_count = data_count
         ...:         self.offset = offset
         ...: 

[ins] In [5]: asizeof.asizeof(MFTRecord("foo", 0, 0))
Out[5]: 136

[ins] In [6]: asizeof.asizeof(("foo", 0, 0))
Out[6]: 144
```

ikelos replied:
Ok, I'm not too keen but... how do dataclasses compare?  Happy to go with whichever works out the smallest... 5:P

ikelos commented on volatility3/framework/plugins/windows/mftscan.py:
```diff
@@ -17,6 +23,22 @@
+
+    def __init__(
+        self,
+        record_name: Optional[str] = None,
```
Allowing all of these to be blank enables the use of `DefaultDict` but strikes me that it could lead to a lot of potential junk entries?

dgmcdona replied:
The use of `DefaultDict` in this effort was just making explicit behavior that was already implicit in the original implementation, but I'll think about possible alternatives.

ikelos commented on volatility3/framework/plugins/windows/mftscan.py:
```diff
@@ -178,7 +201,7 @@ def parse_mft_records(
 
             attr_data = attr.Attr_Data.cast(fn_object)
-            file_name = attr_data.get_full_name()
+            file_name = str(attr_data.get_full_name())
```
I generally try to get people to leave string objects as strings (in case callers need to get back to that information), but sadly for memory reasons it's probably best to allow it here...  5:S  I guess we can't just do this on the generator so they go out of the plugin as strings but are otherwise carried around as the full types?  My concern would be other people calling this (although the type information doesn't imply they come with provenance information so... yeah, probably ok, I'm just trying to figure out how to generally discourage it except in situations of memory exhaustion)

ikelos commented on volatility3/framework/plugins/windows/mftscan.py:
```diff
@@ -236,53 +264,53 @@ def parse_data_record(
-        if mft_record.vol.offset not in record_map:
-            # file name, DATA count, offset
-            record_map[mft_record.vol.offset] = [renderers.NotAvailableValue(), 0, None]
+        rec = record_map[mft_record.vol.offset]
```
Oh, now I see why you used the `DefaultDict`.  Given that `namedtuples` need to be constructed initially anyway (and then can't be changed), I'd sooner build up the component, construct the `namedtuple` and then assign it into the record_map.  I'm guessing this doesn't ever overwrite existing answers (otherwise that's also a conflict we should cope with directly, rather than just blatting over prior results without noting whether there's a discrepancy)...

ikelos commented on volatility3/framework/plugins/windows/mftscan.py:
```diff
@@ -236,53 +264,53 @@ def parse_data_record(
-        record_map: Dict[int, Tuple[str, int, int]],
-        mft_record: interfaces.objects.ObjectInterface,
-        attr: interfaces.objects.ObjectInterface,
+        record_map: DefaultDict[str, MFTRecord],
```
```suggestion
        record_map: Dict[str, MFTRecord],
```
Again, pretty sure type can just be `Dict[str, MFTRecord]`

ikelos commented on volatility3/framework/plugins/windows/mftscan.py:
```diff
@@ -427,11 +455,11 @@ def get_requirements(cls):
-        record_map: Dict[int, Tuple[str, int, int]],
-        mft_record: interfaces.objects.ObjectInterface,
-        attr: interfaces.objects.ObjectInterface,
+        record_map: DefaultDict[str, MFTRecord],
```
```suggestion
        record_map: Dict[str, MFTRecord],
```
Again, the type doesn't need to be DefaultDict, I think just Dict should be fine here?

ikelos commented on volatility3/framework/symbols/windows/extensions/mft.py:
```diff
@@ -10,15 +10,15 @@
     """This represents the base MFT Record"""
 
-    def get_signature(self) -> str:
+    def get_signature(self) -> "objects.String":
```
```suggestion
    def get_signature(self) -> objects.String:
```
Does importing the objects class cause a circular reference here? If not, I'd ditch the quotes, they're usually used to delay requirements to avoid needing to import classes, but it looks like the class has already been imported by now?

ikelos commented on volatility3/framework/symbols/windows/extensions/mft.py:
```diff
@@ -10,15 +10,15 @@
     """This represents an MFT $FILE_NAME Attribute"""
 
-    def get_full_name(self) -> str:
+    def get_full_name(self) -> "objects.String":
```
```suggestion
    def get_full_name(self) -> objects.String:
```

ikelos commented on volatility3/framework/symbols/windows/extensions/mft.py:
```diff
@@ -28,7 +28,7 @@ def get_full_name(self) -> str:
     """This represents an MFT ATTRIBUTE"""
 
-    def get_resident_filename(self) -> Optional[str]:
+    def get_resident_filename(self) -> Optional["objects.String"]:
```
```suggestion
    def get_resident_filename(self) -> Optional[objects.String]:
```

ikelos commented on volatility3/framework/symbols/windows/extensions/mft.py:
```diff
@@ -51,7 +51,7 @@ def get_resident_filename(self) -> Optional[str]:
             return None
 
-    def get_resident_filecontent(self) -> Optional[bytes]:
+    def get_resident_filecontent(self) -> Optional["objects.Bytes"]:
```
```suggestion
    def get_resident_filecontent(self) -> Optional[objects.Bytes]:
```

ikelos commented on volatility3/framework/plugins/windows/mftscan.py:
```diff
@@ -357,7 +385,7 @@ def get_requirements(cls):
     def parse_ads_data_records(
         cls,
-        record_map: Dict[int, Tuple[str, int, int]],
+        record_map: DefaultDict[str, MFTRecord],
```
```suggestion
        record_map: Dict[str, MFTRecord],
```

ikelos commented on volatility3/framework/plugins/windows/mftscan.py:
```diff
@@ -204,11 +227,11 @@ def parse_mft_records(
-        record_map: Dict[int, Tuple[str, int, int]],
+        mft_record: mft.MFTEntry,
+        attr: mft.MFTAttribute,
+        record_map: DefaultDict[str, MFTRecord],
```
```suggestion
        record_map: Dict[str, MFTRecord],
```

ikelos commented on volatility3/framework/plugins/windows/mftscan.py:
```diff
@@ -131,18 +154,18 @@ def enumerate_mft_records(
-        record_map: Dict[int, Tuple[str, int, int]],
-        mft_record: interfaces.objects.ObjectInterface,
-        attr: interfaces.objects.ObjectInterface,
+        record_map: DefaultDict[str, MFTRecord],
```
```suggestion
        record_map: Dict[str, MFTRecord],
```
The type here could surely just be a Dict?

ikelos commented on volatility3/framework/plugins/windows/mftscan.py:
```diff
@@ -53,14 +75,14 @@ def enumerate_mft_records(
-                Dict[int, Tuple[str, int, int]],
-                interfaces.objects.ObjectInterface,
-                interfaces.objects.ObjectInterface,
+                DefaultDict[str, MFTRecord],
```
```suggestion
                Dict[str, MFTRecord],
```

ikelos reviewed (commented):
Generally looks good thanks, a couple quibbles.  I don't really like `DefaultDict` since it leaves you with a dictionary with a whole bunch of potentially bad data (if anyone asks for a random offset from it), so I'd prefer to keep them out of the code.

I also spent a long time trying to get people to leave the volatility information in place when returning (particularly from an API call, rather than a generator), but in this case given the excessive memory it consumes, I'll allow it.  I'd quite like a comment adding next to the casts, to say that it's specifically for memory consumption reasons (to avoid people using it as excuse to do it in their code all the time)...

Lastly type hinting using quotes should only be needed when there's circular/delayed import issues, which I don't think there are here.

Otherwise looks really neat (and hopefully drops the memory usage for this plugin even further!).  5:)

dgmcdona committed:
Remove unnecessary quotes from type hints

Co-authored-by: ikelos <ikelos@users.noreply.github.com>

dgmcdona committed:
Add missing 'e' to variable names

dgmcdona committed:
Create attribute iterator method

Moves logic for iterating through `MFTEntry` attributes into a new
`attributes()` method on the extension class.

dgmcdona committed:
Add attribute iterator to MFTEntry extension class

dgmcdona commented:
Converting back to draft - I think it may be possible to avoid this dict altogether, I'll push some more updates shortly.

dgmcdona committed:
Refactor: Ditch dictionary usage, eliminate callbacks

This simplifies the design of these plugins by moving as much MFTEntry
specific data into the extension class (caching attributes, since
they'll need to be accessed repeatedly) and moving away from the
callback-based implementation to one where classmethods consume
`mft.MFTEntry` objects in order to produce their values.

These changes do two important things:
- They allow us to preserve `object.String` objects until the generator
  function, which makes the public interface much better since people
  can navigate back the the source of the data within their context
- Completely eliminates the `record_map` that was causing so much memory
  consumption.

dgmcdona committed:
Log `InvalidAddressException` instances

dgmcdona commented on volatility3/framework/plugins/windows/mftscan.py:
```diff
@@ -153,150 +145,104 @@
+                    conversion.wintime_to_datetime(std_information.AccessedTime),
+                    renderers.NotApplicableValue(),
+                )
+        except exceptions.InvalidAddressException:
```
fixed

dgmcdona commented:
Looking deeper into the way the plugins were working got me to realize that the dictionary and that class with `__slots__` were not necessary, but it took a good bit of refactoring to remove them. A lot of things have been moved into the extension classes, especially regarding attribute access on `MFTEntry` objects so it's much more convenient to work with them now, although putting that much of what used to be in the plugin logic into the extension classes kind of makes me wish they were versioned. I've heard mention of doing something like that in the past but I don't know if it's feasible right now.

The classmethods now preserve the `objects.String` objects, and they only get converted to regular `str` in the generator methods.

dgmcdona committed:
Major version bumps for all three plugins

ikelos commented:
> The classmethods now preserve the `objects.String` objects, and they only get converted to regular `str` in the generator methods.

Do we still get the same memory savings now?  Also, there is now an example of versioning extensions (not extension classes per-se though).  Check out [the tainting class](https://github.com/volatilityfoundation/volatility3/blob/6ddb0fef2907364283ebc4ddf5c878a50f25fb31/volatility3/framework/symbols/linux/utilities/tainting.py#L9) as an example...

dgmcdona commented:
The only thing I'm really not sure about right now was if it was a good idea to override the default `objects.StructType` constructor for `MFTEntry` in order to pull `symbol_table_name` from `kwargs`. I did it to keep that parameter out of every method signature on the extension class, but that means that they'll all raise `ValueError` if it isn't constructed with it.

dgmcdona commented:
> > The classmethods now preserve the `objects.String` objects, and they only get converted to regular `str` in the generator methods.
> 
> Do we still get the same memory savings now? Also, there is now an example of versioning extensions (not extension classes per-se though). Check out [the tainting class](https://github.com/volatilityfoundation/volatility3/blob/6ddb0fef2907364283ebc4ddf5c878a50f25fb31/volatility3/framework/symbols/linux/utilities/tainting.py#L9) as an example...

We should get even more memory savings, since the dictionary doesn't exist anymore, and `objects.String` is still converted to `str` before being placed in the `TreeNode`.

ikelos commented on volatility3/framework/symbols/windows/extensions/mft.py:
```diff
@@ -2,23 +2,152 @@
     """This represents the base MFT Record"""
 
-    def get_signature(self) -> str:
+    def __init__(
```
I'm not sure I'm keen on this, because it then means anyone following a pointer to an MFTEntry won't actually be able to use it because it won't have a symbol_table_name (even though it will have been pulled from, presumably, the right symbol table)?

dgmcdona committed:
Bump required version numbers

ikelos commented on volatility3/framework/symbols/windows/extensions/mft.py:
```diff
@@ -2,23 +2,152 @@
+    ) -> None:
+        super().__init__(context, type_name, object_info, size, members)
+
+        self._symbol_table_name = kwargs.get("symbol_table_name")
```
Assuming the needed types are in the same table as this type, you can get this information from self.vol.type_name (there's helper functions to get the symbol table part of it, I think)?  I'd be much happier with that than requiring it as an extra argument that can cause exceptions if it hasn't been initialized...

dgmcdona replied:
Awesome thanks! I'll figure out how to get it from there instead.

ikelos commented on volatility3/framework/plugins/windows/mftscan.py:
```diff
@@ -22,7 +20,21 @@ class MFTScan(interfaces.plugins.PluginInterface, timeliner.TimeLinerInterface):
 
     _required_framework_version = (2, 0, 0)
```
```suggestion
    _required_framework_version = (2, 26, 0)
```
This needs bumping because of the extension class changes you're making.

ikelos reviewed (commented):
Feels cleaner, but as you say, loads a lot into the extension class, which I'm fine with, but we don't have a way to version those yet.  I'm also not too keen on the method for getting the symbol table, I'd extract the `vol.type_name` which should include a preceding symbol table name...

ikelos commented:
> The only thing I'm really not sure about right now was if it was a good idea to override the default `objects.StructType` constructor for `MFTEntry` in order to pull `symbol_table_name` from `kwargs`. I did it to keep that parameter out of every method signature on the extension class, but that means that they'll all raise `ValueError` if it isn't constructed with it.

Yeah, you should be able to pull this from `self.vol.type_name` I think...

dgmcdona committed:
Add versioning to MFT extension classes

ikelos commented on volatility3/framework/symbols/windows/extensions/mft.py:
```diff
@@ -5,14 +5,20 @@
 
 
-class MFTEntry(objects.StructType):
+class MFTEntry(objects.StructType, interfaces.configuration.VersionableInterface):
```
I'm not sure this is a good plan, since they're applied to the symbol tables by the framework, you may get handed one and have no idea which version it is (and checking it after the fact will be problematic)...  5:S

dgmcdona replied:
Ok, reverted

dgmcdona committed:
Convert remaining values to Python primitives

dgmcdona committed:
Revert "Add versioning to MFT extension classes"

This reverts commit 1d20e6575908e118ad71746ff9d64b6cd5d23d9f.

dgmcdona commented on volatility3/framework/symbols/windows/extensions/mft.py:
```diff
@@ -2,33 +2,180 @@
 
 
-class MFTAttribute(objects.StructType):
+class MFTAttribute(objects.StructType, interfaces.configuration.VersionableInterface):
```
reverted

dgmcdona committed:
Remove symbol_table_name from object constructor

Get from `self.vol.type_name` instead

dgmcdona commented on volatility3/framework/symbols/windows/extensions/mft.py:
```diff
@@ -2,33 +2,180 @@
 
-class MFTEntry(objects.StructType):
+
+class MFTEntry(objects.StructType, interfaces.configuration.VersionableInterface):
```
resolved.

dgmcdona commented on volatility3/framework/symbols/windows/extensions/mft.py:
```diff
@@ -2,33 +2,180 @@
+                yield attr
+
+
+class MFTFileName(objects.StructType, interfaces.configuration.VersionableInterface):
```
resolved

dgmcdona committed:
Bump required framework version on all three MFTScan plugins

dgmcdona committed:
Fix up remaining type-hints

Updates type hints on some fields of the result namedtuples to be their
`objects.Primitive` types instead of Python primitives, and does any
conversion to Python primitives in the generator methods.

dgmcdona committed:
Rename methods, add docstrings

Improves the naming of a couple of the new extension class methods to
more accurately reflect the return type, and adds docstrings to
extensions class methods.

dgmcdona committed:
Also yield STANDARD_INFORMATION timestamps in timeliner

dgmcdona committed:
Add a filename to STANDARD_INFORMATION timeline entries

ikelos reviewed (approved):
Looks good, thanks very much!  5:D
